Use drag shadows to dragged blocks. Also add ClipDataDescription.

AnmAtAnm opened the pull request:
New DragShadowBuilder renders the dragged block group to an image during dragging.  This improves drag responsiveness and begins the drag view above all over views, even outside of the workspace.

Also, annotate every block drag attempt with ClipDataDescription. A BlockClipDataHelper allows any view to participate in the drag behavior. In a pending CL, the TrashCan will utilize this, but it offers extension to other views outside the Blockly framework (e.g., a sprite, as in Scratch, for copying a block group between workspaces).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/486)
<!-- Reviewable:end -->

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java:
```diff
@@ -381,6 +386,27 @@ public BlockViewFactory onCreateBlockViewFactory(WorkspaceHelper helper) {
+     * <p/>
+     * By default, it constructs a {@link SingleMimeTypeClipDataHelper} with a MIME type derived
+     * from the application's package name. This assumes all Blockly workspaces in an app work with
+     * the same shared set of blocks, and blocks can be dragged/copied/pasted them, even if they are
```
missing across or between in 'dragged/copied/pasted them'

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java:
```diff
@@ -0,0 +1,74 @@
+    boolean isBlockData(ClipDescription clipDescrip);
+
+    /**
+     * Extracts a PendingDrag from a {@link DragEvent}. Assumes the {@link #isBlockData} has been
```
s/the/that

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Override
+    public void setController(BlocklyController controller) {
```
Only the Context is used. Why not just take the Context in the constructor, or take the label instead of the resId and then not even the Context is needed outside the static constructor?

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java:
```diff
@@ -0,0 +1,74 @@
+
+    /**
+     * Constructs a new populated {@link ClipData} using the information from a {@link PendingDrag}.
+     * It also sets the PendingDrag as the backing local state.
```
The implementation doesn't cache the PendingDrag anywhere. Remove this line?

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/ui/Dragger.java:
```diff
@@ -663,6 +708,43 @@ private void updateBlockPosition(DragEvent event) {
+            mSizeY = dragGroup.getHeight();
+        }
+
+        public void onProvideShadowMetrics(Point shadowSize, Point shadowTouchPoint) {
```
Missing @Override annotation

AnmAtAnm replied:
Done.

rachel-fenichel commented:
Reviewed 2 of 15 files at r1.
Review status: 2 of 15 files reviewed at latest revision, 11 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 404 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3k3O9MpNASg2ZA04R:-Ka3k3O9MpNASg2ZA04S:b-phv0u3) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L404)):*
> ```Java
>      */
>     protected BlockClipDataHelper onCreateClipDataHelper() {
>         String packageName = getApplication().getPackageName();
> ```

Why not just call getDefault()? 

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java, line 53 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3iR1b2FAwlOGvGbDG:-Ka3iR1b2FAwlOGvGbDH:b-khe95c) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java#L53)):*
> ```Java
>      *
>      * @param pendingDrag The source of clip
>      * @return A new {@link ClipData} representing the drag and dragged {@link BlockGroup}.
> ```

I don't understand this comment.
"representing the drag and dragged blockGroup"?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java, line 59 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3ixvdX0mJeI6M1q9z:-Ka3ixvevMNjyzVcE_r7:b-uxmgpp) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java#L59)):*
> ```Java
>     /**
>      * This determines whether an incoming {@link ClipData} is a representation of Blockly
>      * {@link Block}s that can be handled by this {@link BlockClipDataHelper}.
> ```

If this just checks if they are blockly blocks, leave of "that can be handled by..."
If this does more checks, I think it should be named isValidBlockData.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java, line 34 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3jLF26qv4W0w3kE3O:-Ka3jLF26qv4W0w3kE3P:bpewk9n) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java#L34)):*
> ```Java
> 
> /**
>  * Implements ClipDataTransformer with a single support MIME type.  Uses intent extras as the
> ```

*supported MIME type
(I think)

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java, line 52 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3jcTWFld7GVuSaCrF:-Ka3jcTWFld7GVuSaCrG:b-bgvh4f) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java#L52)):*
> ```Java
>         String mimeType = "application/x-blockly-" + context.getPackageName() + "+xml";
>         return new SingleMimeTypeClipDataHelper(mimeType, R.string.blockly_clipdata_label_default);
> ```

Nit: delete newline

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java, line 169 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3ksa1S_h7FqvBTaNS:-Ka3ksa2g4qbKXwqezWf:bj9h3fg) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java#L169)):*
> ```Java
>      *
>      * @param pendingDrag The {@link PendingDrag} for the gesture.
>      * @return The {@link BlockGroup} return by {@link OnDragListBlock#getDraggableBlockGroup}.
> ```

Update @return type

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/486)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 11 of 15 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/Dragger.java, line 614 at r2](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka40_C5z0Pzm9rn_x2t:-Ka40_C6D9C-4-rd8ijp:b-1kcsqj) ([raw file](https://github.com/google/blockly-android/blob/858e25950d7b048e7ebfe2194f0637d0f96f7bc0/blocklylib-core/src/main/java/com/google/blockly/android/ui/Dragger.java#L614)):*
> ```Java
>                                     flags);
>                         } catch (IOException e) {
>                             // Serialization failed in ClipDataHelper.
> ```

What would cause serialization to fail?  Is it worth putting a log here or throwing an exception with more information?

---

*[blocklytest/src/androidTest/java/com/google/blockly/android/control/BlocklyControllerTest.java, line 19 at r2](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka417BPwgF8pYi5i7Ms:-Ka417BPwgF8pYi5i7Mt:btcngrt) ([raw file](https://github.com/google/blockly-android/blob/858e25950d7b048e7ebfe2194f0637d0f96f7bc0/blocklytest/src/androidTest/java/com/google/blockly/android/control/BlocklyControllerTest.java#L19)):*
> ```Java
> 
> import com.google.blockly.android.BlocklyTestCase;
> import com.google.blockly.android.clipboard.SingleMimeTypeClipDataHelper;
> ```

Nit: unnecessary import

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/486)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 9 of 14 files reviewed at latest revision, 13 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 395 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-K_pxwlVR-cP1qZYAJkJ-r1-395:-Ka48TCPZwWnN6-5ZLsW:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L395)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

missing across or between in 'dragged/copied/pasted them'
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 404 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3k3O9MpNASg2ZA04R:-Ka48ToPGtt0DgiDduOa:bz8ij48) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L404)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why not just call getDefault()? 
</blockquote></details>

Yeah, I noticed that when fixing Erik's above comment.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java, line 50 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-K_pxwlVR-cP1qZYAJkK-r1-50:-Ka48dy4nlSlUtb18ntf:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java#L50)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

The implementation doesn't cache the PendingDrag anywhere. Remove this line?
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java, line 53 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3iR1b2FAwlOGvGbDG:-Ka48uCqDkBjl835gMXj:bz6x8dy) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java#L53)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I don't understand this comment.
"representing the drag and dragged blockGroup"?
</blockquote></details>

Fixed. (Fail brain.)

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java, line 59 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3ixvdX0mJeI6M1q9z:-Ka49iOeroKx1rhGo3t7:b-1fwgzj) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java#L59)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If this just checks if they are blockly blocks, leave of "that can be handled by..."
If this does more checks, I think it should be named isValidBlockData.
</blockquote></details>

Technically, it can distinguish different types of blocks, but that would be an unusual use case.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java, line 67 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-K_pxwlVR-cP1qZYAJkK-r1-67:-Ka48cvtsrgTav3BoA29:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/BlockClipDataHelper.java#L67)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

s/the/that
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java, line 34 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3jLF26qv4W0w3kE3O:-Ka4A2MnQefTeumqQxMv:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java#L34)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

*supported MIME type
(I think)
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java, line 52 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3jcTWFld7GVuSaCrF:-Ka4A7p1_12Hh3RGkRMG:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java#L52)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Nit: delete newline
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java, line 76 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-K_pxwlVR-cP1qZYAJkL-r1-76:-Ka4AC7GUJ14QoE7fqbW:b28afxy) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/clipboard/SingleMimeTypeClipDataHelper.java#L76)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Only the Context is used. Why not just take the Context in the constructor, or take the label instead of the resId and then not even the Context is needed outside the static constructor?
</blockquote></details>

Cleaned up.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java, line 169 at r1](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka3ksa1S_h7FqvBTaNS:-Ka4Dm-Q3XK04YrQtQ7z:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/e12eeb7f14b1fe2212223545dafc07f7d21dc8a1/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java#L169)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Update @return type
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/Dragger.java, line 614 at r2](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka40_C5z0Pzm9rn_x2t:-Ka4EHv9ShqgKHSrY_FK:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/858e25950d7b048e7ebfe2194f0637d0f96f7bc0/blocklylib-core/src/main/java/com/google/blockly/android/ui/Dragger.java#L614)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What would cause serialization to fail?  Is it worth putting a log here or throwing an exception with more information?
</blockquote></details>

Done.

---

*[blocklytest/src/androidTest/java/com/google/blockly/android/control/BlocklyControllerTest.java, line 19 at r2](https://reviewable.io:443/reviews/google/blockly-android/486#-Ka417BPwgF8pYi5i7Ms:-Ka4Gk1jqJYZX5SvaOwx:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/858e25950d7b048e7ebfe2194f0637d0f96f7bc0/blocklytest/src/androidTest/java/com/google/blockly/android/control/BlocklyControllerTest.java#L19)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Nit: unnecessary import
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/486)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
As discussed, remove setController API as it's not necessary for our use cases. Then <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 9 of 14 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/486#-:-Ka4XM_SWnQDxGiSAMQz:byyjrrn)*
<!-- Sent from Reviewable.io -->
